src/timing.c

```diff
@@ -113,18 +138,123 @@ void jl_destroy_timing(void)
     jl_ptls_t ptls = jl_current_task->ptls;
     jl_timing_block_t *stack = ptls->timing_stack;
     while (stack) {
-        _jl_timing_block_destroy(stack);
+        jl_timing_block_end(stack);
         stack = stack->prev;
     }
 }
 
-jl_timing_block_t *jl_pop_timing_block(jl_timing_block_t *cur_block)
+static const int get_timing_subsystem(const char *subsystem) {
+    const char **match = (const char **)bsearch(
+        &subsystem, jl_timing_subsystems, JL_TIMING_SUBSYSTEM_LAST,
+        sizeof(const char *), indirect_strcmp
+    );
+    if (!match)
+        return JL_TIMING_SUBSYSTEM_LAST;
+
+    return (int)(match - &jl_timing_subsystems[0]);
+}
+
+// This function is analogous to __itt_event_create but for the counts backend
+//
+// `event` is required to live forever
+JL_DLLEXPORT jl_timing_counts_event_t *jl_timing_counts_event_create(const char *event) {
+#ifdef USE_TIMING_COUNTS
+    const size_t n = jl_timing_counts_events.len;
+    for (size_t i = 0; i < n; i++) {
+        jl_timing_counts_event_t *other_event = (jl_timing_counts_event_t *)jl_timing_counts_events.items[i];
+        if (strcmp(event, other_event->name) == 0)
+            return other_event;
+    }
+
+    // No matching event found - create a new one
+    jl_timing_counts_event_t *new_event = (jl_timing_counts_event_t *)malloc(sizeof(jl_timing_counts_event_t));
+    arraylist_push(&jl_timing_counts_events, (void *)new_event);
+    new_event->name = event;
+    jl_atomic_store_relaxed(&new_event->total, 0);
+    return new_event;
+#else
+    return NULL;
+#endif
+}
+
+JL_DLLEXPORT jl_timing_event_t *_jl_timing_event_create(const char *subsystem, const char *name, const char *function, const char *file, int line, int color) {
+    int maybe_subsystem = get_timing_subsystem(subsystem);
+    if (maybe_subsystem >= JL_TIMING_SUBSYSTEM_LAST) {
+        jl_errorf("invalid timing subsystem name: %s", subsystem);
+        return NULL;
+    }
+
+    jl_timing_event_t *event = (jl_timing_event_t *) malloc(sizeof(jl_timing_event_t));
+    event->subsystem = maybe_subsystem;
+
+#ifdef USE_TIMING_COUNTS
+    event->counts_event = jl_timing_counts_event_create(name);
+#endif // USE_TIMING_COUNTS
+
+#ifdef USE_TRACY
+    event->tracy_srcloc.name = name;
+    event->tracy_srcloc.function = function;
+    event->tracy_srcloc.file = file;
+    event->tracy_srcloc.line = line;
+    event->tracy_srcloc.color = color;
+#endif // USE_TRACY
+
+#ifdef USE_ITTAPI
+    event->ittapi_event = __itt_event_create(name, strlen(name));
```

Comment:
I don't know if ITTAPI handles duplicate events properly, so unless it says something about that I would like to have some deduplication logic here.